fx quant: enable qconfig_dict to target function invocations by order

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59605 fx quant: enable qconfig_dict to target function invocations by order**

Summary:

Enables targeting of individual function invocations by execution order.
For example, given a module such as

```
class M1(torch.nn.Module):
  def forward(self, x):
    x = torch.add(x, x)
    x = torch.add(x, x)
    return x

class M2(torch.nn.Module):
  def __init__(self):
    self.m1 = M1()

  def forward(self, x):
    x = self.m1(x)
    return x
```

We can now target the first add of `m1` with

```
qconfig_dict = {
  "module_name_object_type_order": ("m1", torch.add, 0, custom_qconfig),
}
```

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_qconfig_module_name_object_type_order
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28951077](https://our.internmc.facebook.com/intern/diff/D28951077)

vkuzo committed:
fx quant: enable qconfig_dict to target function invocations by order

Summary:

Enables targeting of individual function invocations by execution order.
For example, given a module such as

```
class M1(torch.nn.Module):
  def forward(self, x):
    x = torch.add(x, x)
    x = torch.add(x, x)
    return x

class M1(torch.nn.Module):
  def __init__(self):
    self.m1 = M1()

  def forward(self, x):
    x = self.m1(x)
    return x
```

We can now target the first add of `m1` with

```
qconfig_dict = {
  "module_name_function_order": ("m1", torch.add, 0, custom_qconfig),
}
```

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_qconfig_module_name_function_order
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28951077).

hx89 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -1033,6 +1033,67 @@ def forward(self, x):
+                return x
+
+        m = M3().eval()
+        qconfig_dict = {
```
A common use case is that users want to specify a few layers to skip quantization in qconfig_dict, while most other layers are kept quantized using some global qconfig, in which case it's more convenient to specify layers to skip quantization in qconfig_dict instead of layers to quantize. How could users setup qconfig_dict for such use case?

vkuzo replied:
one could specify

```
qconfig_dict = {
  "module_name_function_order": ("m1", torch.add, 0, None),
}
```

to skip quantization of the first add, and so forth

hx89 reviewed (approved):
Thanks for adding this great feature!

jerryzh168 reviewed (approved):
lgtm, thanks!

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -1033,6 +1033,67 @@ def forward(self, x):
         self.assertEqual(m.module_conv1.qconfig, module_name_regex_qconfig)
         self.assertEqual(m.module_conv2.qconfig, module_name_qconfig)
 
+    def test_qconfig_module_name_function_order(self):
```
can we also test the precedence of this config?

jerryzh168 commented on torch/quantization/fx/qconfig_utils.py:
```diff
@@ -86,6 +86,27 @@ def get_module_name_regex_qconfig(qconfig_dict, module_name, fallback_qconfig):
     return fallback_qconfig
 
 
+def maybe_adjust_qconfig_for_module_name_function_order(
```
should we merge this function with get_object_type_qconfig to make the precedence more clear?

vkuzo replied:
yes, I looked into this, it's not clean with the current code.  We might need to do some refactoring of the qconfig utils to make this easier, in a future PR.

jerryzh168 commented on torch/quantization/fx/qconfig_utils.py:
```diff
@@ -134,6 +165,14 @@ def generate_qconfig_map(
             module_path, module_type = node_name_to_scope[node.name]
             qconfig = get_qconfig(
                 qconfig_dict, module_type, module_path, function_qconfig)
+
```
I think the precedence should probably be:
function_order_config > global_function_type_config > module_scope_config > global_config

jerryzh168 replied:
where module_scope_config is from module_name_config or global module_type_config and the precedence there is module_name_config > module_type_config (this is taken care of by get_qconfig function I think)

vkuzo replied:
> I think the precedence should probably be: function_order_config > global_function_type_config > module_scope_config > global_config

yes, this is true, although not the easiest to see with the current code.  Saving cleaning up the qconfig tools to make the logic easier to understand for a future PR, hope that's ok!

vkuzo committed:
Update on "fx quant: enable qconfig_dict to target function invocations by order"


Summary:

Enables targeting of individual function invocations by execution order.
For example, given a module such as

```
class M1(torch.nn.Module):
  def forward(self, x):
    x = torch.add(x, x)
    x = torch.add(x, x)
    return x

class M2(torch.nn.Module):
  def __init__(self):
    self.m1 = M1()

  def forward(self, x):
    x = self.m1(x)
    return x
```

We can now target the first add of `m1` with

```
qconfig_dict = {
  "module_name_function_order": ("m1", torch.add, 0, custom_qconfig),
}
```

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_qconfig_module_name_function_order
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28951077](https://our.internmc.facebook.com/intern/diff/D28951077)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28951077).

jerryzh168 commented on torch/quantization/quantize_fx.py:
```diff
@@ -289,6 +289,15 @@ def prepare_fx(
+        ("foo.bar", torch.nn.functional.linear, 0, qconfig?),
+      ],
+
       # priority (in increasing order): global, object_type, module_name_regex, module_name
```
nit: please also remember to include the precedence in this list as well, we probably want to have two precedence list: one for module and one for function to make it more clear. (can be done in a separate PR)

vkuzo replied:
thanks, will fix this

jerryzh168 reviewed (approved):
sg, thanks, we can do refactor in a separate PR.

vkuzo committed:
Update on "fx quant: enable qconfig_dict to target function invocations by order"


Summary:

Enables targeting of individual function invocations by execution order.
For example, given a module such as

```
class M1(torch.nn.Module):
  def forward(self, x):
    x = torch.add(x, x)
    x = torch.add(x, x)
    return x

class M2(torch.nn.Module):
  def __init__(self):
    self.m1 = M1()

  def forward(self, x):
    x = self.m1(x)
    return x
```

We can now target the first add of `m1` with

```
qconfig_dict = {
  "module_name_function_order": ("m1", torch.add, 0, custom_qconfig),
}
```

Test Plan:

```
python test/test_quantization.py TestQuantizeFx.test_qconfig_module_name_function_order
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28951077](https://our.internmc.facebook.com/intern/diff/D28951077)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28951077).

jerryzh168 commented on torch/quantization/fx/qconfig_utils.py:
```diff
@@ -172,7 +228,9 @@ def check_is_valid_qconfig_dict(qconfig_dict: Any) -> None:
-    qconfig_dict_allowed_keys = {"", "object_type", "module_name_regex", "module_name"}
+    qconfig_dict_allowed_keys = {
+        "", "object_type", "module_name_regex", "module_name",
+        "module_name_object_type_order"}
```
nit: should this just be `object_type_order`?

vkuzo replied:
hmm, it is `object_type_order` within the forward function of `module_name`.  Thoughts on if there is a better way to name it?

jerryzh168 reviewed (approved):
lgtm
